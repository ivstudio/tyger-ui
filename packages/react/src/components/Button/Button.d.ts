export type ButtonVariant = 'filled' | 'outlined' | 'text';
export type ButtonSizes = 'sm' | 'md' | 'lg';

export interface ButtonProps {
    id?: string;
    className?: string;
    type?: 'submit' | 'button' | 'reset';
    disabled?: boolean;
    size?: ButtonSizes;
    fullWidth?: boolean;
    children?: React.ReactNode;
    variant?: ButtonVariant;
    rounded?: boolean;
    startIcon?: React.ReactNode;
    endIcon?: React.ReactNode;
    onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export type ButtonRootProps = {
    size: ButtonSizes;
    fullWidth: boolean;
    variant: ButtonVariant;
    rounded: boolean;
    hasStartIcon?: boolean;
    hasEndIcon?: boolean;
};
