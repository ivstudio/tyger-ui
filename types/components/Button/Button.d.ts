import * as React from 'react';
export declare type TButtonVariant = 'filled' | 'outlined' | 'text';
export declare type TSize = 'sm' | 'md' | 'lg';
export interface IButton {
    id?: string;
    className?: string;
    type?: 'submit' | 'button' | 'reset';
    disabled?: boolean;
    size?: TSize;
    fullWidth?: boolean;
    children?: React.ReactNode;
    variant?: TButtonVariant;
    rounded?: boolean;
    startIcon?: React.ReactNode;
    endIcon?: React.ReactNode;
    onClick(event: React.MouseEvent<HTMLButtonElement>): void;
}
declare const Button: ({ id, children, type, disabled, size, className, fullWidth, rounded, variant, startIcon, endIcon, ...baseProps }: IButton) => JSX.Element;
export default Button;
