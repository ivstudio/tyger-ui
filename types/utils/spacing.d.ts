export declare const spacing: {
    readonly '0': "0px";
    readonly '2': "2px";
    readonly '4': "4px";
    readonly '8': "8px";
    readonly '12': "12px";
    readonly '16': "16px";
    readonly '24': "24px";
    readonly '32': "32px";
    readonly '40': "40px";
    readonly '48': "48px";
    readonly '64': "64px";
    readonly '80': "80px";
    readonly '96': "96px";
    readonly '160': "160px";
};
export declare type TSpacing = keyof typeof spacing;
