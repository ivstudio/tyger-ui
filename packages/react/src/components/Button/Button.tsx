'use client';
import * as React from 'react';
import styled from 'styled-components';

import {
    backgroundTransition,
    buttonBase,
    flexCenterInline,
} from '../../styles';
import Box from '../Box';
import { ButtonProps, ButtonRootProps } from './Button.types';

const buttonSpacing = {
    text: {
        base: {
            sm: `4px 5px`,
            md: `6px 8px`,
            lg: `8px 11px`,
        },
        startIcon: {
            sm: `4px 5px 4px 2px`,
            md: `6px 8px 6px 5px`,
            lg: `8px 11px 8px 6px`,
        },
        endIcon: {
            sm: `4px 2px 4px 5px`,
            md: `6px 5px 6px 8px`,
            lg: `8px 6px 8px 11px`,
        },
    },
    outline: {
        base: {
            sm: `3px 9px`,
            md: `5px 15px`,
            lg: `7px 21px`,
        },
        startIcon: {
            sm: `3px 9px 3px 6px`,
            md: `5px 15px 5px 12px`,
            lg: `7px 21px 7px 16px`,
        },
        endIcon: {
            sm: `3px 6px 3px 9px`,
            md: `5px 12px 5px 15px`,
            lg: `7px 16px 7px 21px`,
        },
    },
    filled: {
        base: {
            sm: `4px 10px`,
            md: `6px 16px`,
            lg: `8px 22px`,
        },
        startIcon: {
            sm: `4px 10px 4px 6px`,
            md: `6px 16px 6px 12px`,
            lg: `8px 22px 8px 16px`,
        },
        endIcon: {
            sm: `4px 6px 4px 10px`,
            md: `6px 12px 6px 16px`,
            lg: `8px 16px 8px 22px`,
        },
    },
} as const;

const fonButtonSizess = {
    sm: {
        fonButtonSizes: '0.8125rem',
        '& svg': {
            fonButtonSizes: '16px',
        },
    },

    md: {
        fonButtonSizes: '0.875rem',
        '& svg': {
            fonButtonSizes: '20px',
        },
    },

    lg: {
        fonButtonSizes: '0.9375rem',
        '& svg': {
            fonButtonSizes: '24px',
        },
    },
} as const;

const roundedCorners = {
    sm: '18px',
    md: '18px',
    lg: '22px',
} as const;

const iconSpacing = {
    sm: '4px',
    md: '4px',
    lg: '6px',
} as const;

const ButtonRoot = styled.button<ButtonRootProps>`
    ${flexCenterInline};
    ${buttonBase};
    ${backgroundTransition};
    min-width: 64px;
    font-weight: 500;
    letter-spacing: 0.02857em;
    ${({ size }) => fonButtonSizess[size]};
    line-height: ${({ theme }) => theme.lineHeight.xl};
    width: ${({ fullWidth }) => fullWidth && '100%'};
    border-radius: ${({ rounded, size }) => {
        return rounded ? roundedCorners[size] : '4px';
    }};
    ${({ theme, variant, size, hasStartIcon, hasEndIcon }) => {
        if (!theme || !variant) return null;
        const { button } = theme;
        const spacingType = hasStartIcon
            ? 'startIcon'
            : hasEndIcon
            ? 'endIcon'
            : 'base';

        switch (variant) {
            case 'text':
                return {
                    color: button.text.color,
                    background: button.text.backgroundColor,
                    padding: buttonSpacing.text[spacingType][size],
                    '&:hover': {
                        background: button.text.hover.backgroundColor,
                    },
                    '&:disabled': {
                        color: button.text.disabled.color,
                        background: button.text.backgroundColor,
                    },
                };
            case 'outlined':
                return {
                    color: button.outlined.color,
                    background: button.outlined.backgroundColor,
                    padding: buttonSpacing.outline[spacingType][size],
                    border: `1px solid ${button.outlined.border}`,
                    '&:hover': {
                        background: button.outlined.hover.backgroundColor,
                        border: `1px solid ${button.outlined.hover.border}`,
                    },
                    '&:disabled': {
                        color: button.outlined.disabled.color,
                        background: button.outlined.disabled.backgroundColor,
                        border: `1px solid ${button.outlined.disabled.border}`,
                    },
                };
            case 'filled':
            default:
                return {
                    color: button.filled.color,
                    background: button.filled.backgroundColor,
                    padding: buttonSpacing.filled[spacingType][size],
                    '&:hover': {
                        background: button.filled.hover.backgroundColor,
                    },
                    '&:disabled': {
                        color: button.filled.disabled.color,
                        background: button.filled.disabled.backgroundColor,
                    },
                };
        }
    }}
`;

const Button = ({
    id,
    children,
    type = 'button',
    disabled = false,
    size = 'md',
    className,
    fullWidth = false,
    rounded = true,
    variant = 'filled',
    startIcon = null,
    endIcon = null,
    ...baseProps
}: ButtonProps) => (
    <ButtonRoot
        {...baseProps}
        id={id}
        type={type}
        size={size}
        disabled={disabled}
        className={className}
        fullWidth={fullWidth}
        variant={variant}
        rounded={rounded}
        hasStartIcon={startIcon !== null}
        hasEndIcon={endIcon !== null}
    >
        {startIcon && !endIcon && (
            <Box
                tag="span"
                styles={{ marginRight: children ? iconSpacing[size] : '0' }}
                flexCenterInline
            >
                {startIcon}
            </Box>
        )}
        {children && children}
        {endIcon && !startIcon && (
            <Box
                tag="span"
                styles={{ marginLeft: children ? iconSpacing[size] : '0' }}
                flexCenterInline
            >
                {endIcon}
            </Box>
        )}
    </ButtonRoot>
);

export default Button;
