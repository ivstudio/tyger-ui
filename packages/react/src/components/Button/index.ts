'use client';
export { default } from './Button';
