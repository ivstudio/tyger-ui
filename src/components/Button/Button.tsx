import * as React from 'react';
import styled from 'styled-components';

import {
    backgroundTransition,
    buttonBase,
    flexCenterInline,
} from '../../styles';
import Box from '../Box';

export type TButtonVariant = 'filled' | 'outlined' | 'text';
export type TSize = 'sm' | 'md' | 'lg';

export interface IButton {
    id?: string;
    className?: string;
    type?: 'submit' | 'button' | 'reset';
    disabled?: boolean;
    size?: TSize;
    fullWidth?: boolean;
    children?: React.ReactNode;
    variant?: TButtonVariant;
    rounded?: boolean;
    startIcon?: React.ReactNode;
    endIcon?: React.ReactNode;
    onClick(event: React.MouseEvent<HTMLButtonElement>): void;
}

const fontSizes = {
    sm: {
        fontSize: '0.8125rem',
        '& svg': {
            fontSize: '16px',
        },
    },

    md: {
        fontSize: '0.875rem',
        '& svg': {
            fontSize: '20px',
        },
    },

    lg: {
        fontSize: '0.9375rem',
        '& svg': {
            fontSize: '24px',
        },
    },
} as const;

const buttonSpacing = {
    text: {
        sm: `4px 5px`,
        md: `6px 8px`,
        lg: `8px 11px`,
    },
    outline: {
        sm: `3px 9px`,
        md: `5px 15px`,
        lg: `7px 21px`,
    },
    filled: {
        sm: `4px 10px`,
        md: `6px 16px`,
        lg: `8px 22px`,
    },
} as const;

const roundedCorners = {
    sm: '14px',
    md: '18px',
    lg: '22px',
} as const;

const iconSpacing = {
    sm: '4',
    md: '4',
    lg: '8',
} as const;

const ButtonRoot = styled.button<{
    size: TSize;
    fullWidth: boolean;
    variant: TButtonVariant;
    rounded: boolean;
}>`
    ${flexCenterInline};
    ${buttonBase};
    ${backgroundTransition};
    min-width: 64px;
    font-weight: 500;
    letter-spacing: 0.02857em;
    ${({ size }) => fontSizes[size]};
    line-height: ${({ theme }) => theme.lineHeight.xl};
    width: ${({ fullWidth }) => fullWidth && '100%'};
    border-radius: ${({ rounded, size }) => {
        return rounded ? roundedCorners[size] : '4px';
    }};
    ${({ theme, variant, size }) => {
        if (!theme || !variant) return null;
        const { button } = theme.palette;
        switch (variant) {
            case 'text':
                return {
                    color: button.text.color,
                    background: button.text.backgroundColor,
                    padding: buttonSpacing.text[size],
                    '&:hover': {
                        background: button.text.hover.backgroundColor,
                    },
                    '&:disabled': {
                        color: button.text.disabled.color,
                        background: button.text.backgroundColor,
                    },
                };
            case 'outlined':
                return {
                    color: button.outlined.color,
                    background: button.outlined.backgroundColor,
                    padding: buttonSpacing.outline[size],
                    border: `1px solid ${button.outlined.border}`,
                    '&:hover': {
                        background: button.outlined.hover.backgroundColor,
                        border: `1px solid ${button.outlined.hover.border}`,
                    },
                    '&:disabled': {
                        color: button.outlined.disabled.color,
                        background: button.outlined.disabled.backgroundColor,
                        border: `1px solid ${button.outlined.disabled.border}`,
                    },
                };
            case 'filled':
            default:
                return {
                    color: button.filled.color,
                    background: button.filled.backgroundColor,
                    padding: buttonSpacing.filled[size],
                    '&:hover': {
                        background: button.filled.hover.backgroundColor,
                    },
                    '&:disabled': {
                        color: button.filled.disabled.color,
                        background: button.filled.disabled.backgroundColor,
                    },
                };
        }
    }}
`;

const Button = ({
    id,
    children,
    type = 'button',
    disabled = false,
    size = 'md',
    className,
    fullWidth = false,
    rounded = true,
    variant = 'filled',
    startIcon = null,
    endIcon = null,
    ...baseProps
}: IButton) => (
    <ButtonRoot
        {...baseProps}
        id={id}
        type={type}
        size={size}
        disabled={disabled}
        className={className}
        fullWidth={fullWidth}
        variant={variant}
        rounded={rounded}
    >
        {startIcon && (
            <Box
                tag="span"
                mr={children ? iconSpacing[size] : '0'}
                flexCenterInline
            >
                {startIcon}
            </Box>
        )}
        {children && children}
        {endIcon && (
            <Box
                tag="span"
                ml={children ? iconSpacing[size] : '0'}
                flexCenterInline
            >
                {endIcon}
            </Box>
        )}
    </ButtonRoot>
);

export default Button;
