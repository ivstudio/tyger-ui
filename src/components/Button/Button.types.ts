export type TButtonVariant = 'primary' | 'outlined' | 'text';

export interface IButton {
    id?: string;
    className?: string;
    type?: 'submit' | 'button' | 'reset';
    disabled?: boolean;
    size?: TSize;
    fullWidth?: boolean;
    children: React.ReactNode | React.ReactNode[];
    variant?: TButtonVariant;
    rounded?: boolean;
    onClick(event: React.MouseEvent<HTMLButtonElement>): void;
}

export type TSize = 'sm' | 'md' | 'lg';
