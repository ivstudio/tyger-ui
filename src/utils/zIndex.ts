export const zindex = {
    mobileStepper: 1000,
    appBar: 1100,
    drawer: 1200,
    modal: 1300,
    snackbar: 1400,
    tooltip: 1500,
};
export type TZIndex = keyof typeof zindex;
