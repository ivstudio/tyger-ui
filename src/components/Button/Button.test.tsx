import React from 'react';
import Button from './Button';
import { render } from '../../../test/utils/mockProvider';

const clickHandler = jest.fn();

describe('Button', () => {
    it('renders Button', () => {
        const { container } = render(
            <Button onClick={clickHandler}>Button</Button>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
