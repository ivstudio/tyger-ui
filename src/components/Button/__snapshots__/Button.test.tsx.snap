// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders Button 1`] = `
<button
  class="sc-gsnTZi gWoXO"
  type="button"
>
  Button
</button>
`;
